### ISSUE TITLE
BART generation results for top k differ in each run

### ISSUE THREAD

**Issue by tuhinjubcse** [pytorch/fairseq#1947]
## 🐛 Bug

Every time I do Bart.sample for topk the results are different. For reproducibility is there a way to fix this?

`                hypotheses_batch = bart.sample(slines, sampling=True, sampling_topk=5, temperature=0.7, lenpen=2.0, max_len_b=30, min_len=7, no_repeat_ngram_size=3)
`
### Environment

Environment
- fairseq Version (e.g., 1.0 or master): 0.9.0
- PyTorch Version (e.g., 1.0) 1.3.1
- OS (e.g., Linux): Linux
- How you installed fairseq (pip, source):
- Build command you used (if compiling from source):
- Python version: 3.7.3
- CUDA/cuDNN version: 10.1.243
- GPU models and configuration: RTX 2080 ( 4 gpus)

**Comment by tuhinjubcse** [comment]
@myleott

**Comment by erip** [comment]
Did you run `bart.eval()` first?

**Comment by tuhinjubcse** [comment]
Yes I did put bart.eval()

```
bart = BARTModel.from_pretrained('checkpoint-similenew/',checkpoint_file='checkpoint_best.pt',data_name_or_path='similenew')

bart.cuda()
bart.eval()
#bart.half()
count = 1
bsz = 128
print("done")
t = 0.7
for val in [5]:
    with open('similenew/test.source') as source, open('similenew/testtopk5_0.7.hypo', 'w') as fout:
        sline = source.readline().strip()
        slines = [sline]
        for sline in source:
            if count % bsz == 0:
                with torch.no_grad():
                    hypotheses_batch = bart.sample(slines, sampling=True, sampling_topk=val  ,temperature=t ,lenpen=2.0, max_len_b=30, min_len=7, no_repeat_ngram_size=3)
                for hypothesis in hypotheses_batch:
                    fout.write(hypothesis.replace('\n','') + '\n')
                    fout.flush()
                slines = []

            slines.append(sline.strip())
            count += 1
        if slines != []:
            hypotheses_batch = bart.sample(slines, sampling=True,   sampling_topk=val  ,temperature=t ,lenpen=2.0, max_len_b=30, min_len=7, no_repeat_ngram_size=3)
            for hypothesis in hypotheses_batch:
                fout.write(hypothesis.replace('\n','') + '\n')
                fout.flush()

```

**Comment by erip** [comment]
I'm a bit confused by your code. You're saying that if you run this script twice, the output file contains different hypotheses?

**Comment by tuhinjubcse** [comment]
```
import torch
from fairseq.models.bart import BARTModel

bart = BARTModel.from_pretrained(
    'checkpoints/',
    checkpoint_file='checkpoint_best.pt',
    data_name_or_path='cnn_dm-bin'
)

bart.cuda()
bart.eval()

Nothing confusing its the same as fairseq repo 
Yes I get different hypothesis in each run
bart.half()
count = 1
bsz = 32
with open('cnn_dm/test.source') as source, open('cnn_dm/test.hypo', 'w') as fout:
    sline = source.readline().strip()
    slines = [sline]
    for sline in source:
        if count % bsz == 0:
            with torch.no_grad():
                hypotheses_batch = bart.sample(slines, beam=4, lenpen=2.0, max_len_b=140, min_len=55, no_repeat_ngram_size=3)

            for hypothesis in hypotheses_batch:
                fout.write(hypothesis + '\n')
                fout.flush()
            slines = []

        slines.append(sline.strip())
        count += 1
    if slines != []:
        hypotheses_batch = bart.sample(slines, beam=4, lenpen=2.0, max_len_b=140, min_len=55, no_repeat_ngram_size=3)
        for hypothesis in hypotheses_batch:
            fout.write(hypothesis + '\n')
            fout.flush()
```

**Comment by erip** [comment]
This is with a pre-trained model or a fine-tuned model? I wasn't able to reproduce. Details below:

----

```sh
$ cat test.source
hello
my
name
is
erip
(fairseq-dev)
~/Downloads/scratch ⌚ 14:36:59
$ cat test.py
import torch

bart = torch.hub.load('pytorch/fairseq', 'bart.large.cnn')

bart.eval()

count = 1
bsz = 32
with open('test.source') as source, open('test.hypo2', 'w') as fout: # changed from 'test.hypo' to 'test.hypo2' between runs
    sline = source.readline().strip()
    slines = [sline]
    for sline in source:
        if count % bsz == 0:
            with torch.no_grad():
                hypotheses_batch = bart.sample(slines, beam=4, lenpen=2.0, max_len_b=140, min_len=55, no_repeat_ngram_size=3)

            for hypothesis in hypotheses_batch:
                fout.write(hypothesis + '\n')
                fout.flush()
            slines = []

        slines.append(sline.strip())
        count += 1
    if slines != []:
        hypotheses_batch = bart.sample(slines, beam=4, lenpen=2.0, max_len_b=140, min_len=55, no_repeat_ngram_size=3)
        for hypothesis in hypotheses_batch:
            fout.write(hypothesis + '\n')
            fout.flush()
(fairseq-dev)
~/Downloads/scratch ⌚ 14:37:14
$ diff test.hypo test.hypo2 | wc -l
       0
```

**Comment by tuhinjubcse** [comment]
fine-tuned model
Also, why do you have one-word sentences  ??

**Comment by erip** [comment]
Shouldn't matter from a reproducibility standpoint. I just created some junk data.

**Comment by tuhinjubcse** [comment]
How are you reproducing ? 
I see you do not use sampling but beam reproduce with my code ?

**Comment by erip** [comment]
Fix your random seed at the beginning of your script:

```python
import torch
import numpy as np

bart = torch.hub.load('pytorch/fairseq', 'bart.large.cnn')

bart.eval()

torch.manual_seed(0)
np.random.seed(0)


count = 1
bsz = 32
with open('test.source') as source, open('test.hypo2', 'w') as fout:
    sline = source.readline().strip()
    slines = [sline]
    for sline in source:
        if count % bsz == 0:
            with torch.no_grad():
                hypotheses_batch = bart.sample(slines, sampling=True, beam=4, lenpen=2.0, max_len_b=140, min_len=55, no_repeat_ngram_size=3)

            for hypothesis in hypotheses_batch:
                fout.write(hypothesis + '\n')
                fout.flush()
            slines = []

        slines.append(sline.strip())
        count += 1
    if slines != []:
        hypotheses_batch = bart.sample(slines, beam=4, sampling=True, lenpen=2.0, max_len_b=140, min_len=55, no_repeat_ngram_size=3)
        for hypothesis in hypotheses_batch:
            fout.write(hypothesis + '\n')
            fout.flush()
```

**Comment by tuhinjubcse** [comment]
Resolved thanks
